lisa/util/shell.py

```diff
@@ -43,14 +46,136 @@ def close(self) -> None:
+        else:
+            assert isinstance(path, Path)
+            exists = path.exists()
+        return exists
+
+    def remove(self, path: PurePath, recursive: bool = False) -> None:
+        self.initialize()
+        if self.isRemote:
+            assert self.innerShell
+            path_str = self._purePathToStr(path)
+            self.innerShell.remove(path_str, recursive)
+        else:
+            assert isinstance(path, Path)
+            path.rmdir()
+
+    def chmod(self, path: PurePath, mode: int) -> None:
+        self.initialize()
+        if self.isRemote:
+            assert self.innerShell
+            path_str = self._purePathToStr(path)
+            self.innerShell.chmod(path_str, mode)
+        else:
+            assert isinstance(path, Path)
+            path.chmod(mode)
+
+    def stat(self, path: PurePath) -> os.stat_result:
+        self.initialize()
+        if self.isRemote:
+            assert self.innerShell
+            path_str = self._purePathToStr(path)
+            sftp_attributes: paramiko.SFTPAttributes = self.innerShell.stat(path_str)
+
+            result = os.stat_result(sftp_attributes.st_mode)
+            result.st_mode = sftp_attributes.st_mode
+            result.st_size = sftp_attributes.st_size
+            result.st_uid = sftp_attributes.st_uid
+            result.st_gid = sftp_attributes.st_gid
+            result.st_atime = sftp_attributes.st_atime
+            result.st_mtime = sftp_attributes.st_mtime
+        else:
+            assert isinstance(path, Path)
+            result = path.stat()
+        return result
+
+    def is_dir(self, path: PurePath) -> bool:
+        self.initialize()
+        if self.isRemote:
+            assert self.innerShell
+            path_str = self._purePathToStr(path)
+            result: bool = self.innerShell.is_dir(path_str)
+        else:
+            assert isinstance(path, Path)
+            result = path.is_dir()
+        return result
+
+    def is_symlink(self, path: PurePath) -> bool:
+        self.initialize()
+        if self.isRemote:
+            assert self.innerShell
+            path_str = self._purePathToStr(path)
+            result: bool = self.innerShell.is_symlink(path_str)
+        else:
+            assert isinstance(path, Path)
+            result = path.is_symlink()
+        return result
+
+    def symlink(self, source: PurePath, destination: PurePath) -> None:
+        self.initialize()
+        if self.isRemote:
+            assert self.innerShell
+            source_str = self._purePathToStr(source)
+            destination_str = self._purePathToStr(destination)
+            self.innerShell.symlink(source_str, destination_str)
+        else:
+            assert isinstance(source, Path)
+            assert isinstance(destination, Path)
+            source.symlink_to(destination)
+
+    def chown(self, path: PurePath, uid: int, gid: int) -> None:
+        self.initialize()
+        if self.isRemote:
```

Comment:
All these conditionals really tell me we want to better abstract shell/node. In my experience with cross-platform support (PowerShell, Mesos, etc.), the most difficult to maintain code was where this pattern was followed, in contrast having a defined interface, and then separately implementing the different scenarios.

Reply:
It definitely can be improved, feel free to improve it.

Reply:
Unfortunately, it'll be harder to do so now that there is code depending on it. We should be thinking about these things before we write the initial implementation.